Add AdGear bidder adaptor

bmilekic opened the pull request:
Bidder adaptor for the AdGear ad platform (http://www.adgear.com/)

bmilekic commented:
@protonate I see you're assigned to review. Let me know if you have any questions or suggestions for improvement!

mkendall07 commented on integrationExamples/gpt/pbjs_example_gpt.html:
```diff
@@ -104,6 +104,13 @@
                     },
                     {
+                        bidder: 'adgear',
+                        // optional params: [ 'url', 'sizes' ]
```
Would you be able to provide examples of `url` and `sizes` ? Is there a reason to have `sizes` optionally here vs level up on the adUnit?

bmilekic replied:
I added some explanation in the comment for url and sizes. The url param is used to optionally override the default window.location.href as the 'declared url', typically for pubs who want to hide that from the ad server. The sizes param is used to optionally override the sizes in the adUnit -- typically used for pubs who have a setup where the adUnit asks for two compatible formats but for whatever reason in AdGear they've only got one of the two defined (e.g., just call for 300x250 instead of both 300x250 and 300x600 defined in the adUnit).

mkendall07 commented on src/adapters/adgear.js:
```diff
@@ -0,0 +1,130 @@
+var Adapter = require('./adapter.js');
+
+var AdGearAdapter = function AdGearAdapter() {
+  var proto = ('http:' === document.location.protocol) ? 'http' : 'https';
```
We are requiring all adapters to hardcode to `HTTPS` now. Please remove the `HTTP` option.

mkendall07 commented on src/adapters/adgear.js:
```diff
@@ -0,0 +1,130 @@
+      }
+    });
+
+    bidmanager.setExpectedBidsCount(bidderCode, bids.length);
```
No need for this line as we automatically set bid count to `bids.length` as a default.

mkendall07 commented on src/adapters/adgear.js:
```diff
@@ -0,0 +1,130 @@
+      if (bidStash) {
+
+        var bidObj = bidStash.bidObj;
+        bidObj.status = CONSTANTS.STATUS.GOOD;
```
This line isn't required anymore. You can use it for creating the bid like so: `bid = bidfactory.createBid(CONSTANTS.STATUS.GOOD);`

mkendall07 commented on src/adapters/adgear.js:
```diff
@@ -0,0 +1,130 @@
+
+        } else {
+
+          bid = bidfactory.createBid(2);
```
Please use `bidfactory.createBid(CONSTANTS.STATUS.NO_BID)`

mkendall07 commented:
@bmilekic thanks for the PR. I added a few comments. Can you please post test params to validate a bid response? If you don't want to post them here you can email us directly.

bmilekic commented:
@mkendall07 Thanks for the review. I believe I've addressed the comments. I rebased off of your master and squashed all commits for diff readability. If in the future you prefer that I leave the PR review amendments as separate commits please LMK.

I'll Email you some sample tagIds for testing.

bmilekic commented:
Hey folks. Any update on this PR?

protonate commented:
We've been working on a maintenance and bug fix release, about to go out as Prebid v0.9.2, and this PR is slated for our next release v0.10.0, expected to go out in about two weeks.

protonate commented on src/adapters/adgear.js:
```diff
@@ -0,0 +1,125 @@
+
+          var cuid = utils.getUniqueIdentifierStr();
+          url = utils.tryAppendQueryString(url, 'AG_CBID', encodeURIComponent(cuid));
+          bidmanager.pbCallbackMap[cuid] = {
```
This is now throwing an error as `pbCallbackMap` has been deprecated. I apologize -- this changed since you first opened this PR. We need some way to match the request with the response, and we can do that by a) returning the bid request ID in your server response, b) use a `createHandleAdGearCallback` function that returns the `handleAdGearCallback` function, forming a closure over which we can include the bidRequest, c) find a request for a given bidder code, which assumes that there is only one such request, or d) another method you prefer.

protonate commented:
Is there an update for this? We're planning to include in next release if we can get this merged soon.

mkendall07 commented:
@bmilekic ^

bmilekic commented:
Hey folks, just caught this. Will dig in today and get back to you.

bmilekic commented:
@mkendall07 Before I refresh the PR, would this approach be OK with you? It keeps close to the original implementation, by basically keeping a global stash keyed by the generated UniqueIdentifierStr within the AdGearAdapter instance itself, to replace the now deprecated global pbCallbackMap (I still have to test and update the PR so LMK if you're OK with this approach and I'll do that if yes):

`diff --git a/src/adapters/adgear.js b/src/adapters/adgear.js
index 76d464a..1c4fbbc 100644
--- a/src/adapters/adgear.js
+++ b/src/adapters/adgear.js
@@ -8,6 +8,7 @@ var Adapter = require('./adapter.js');
 var AdGearAdapter = function AdGearAdapter() {
   var htag = 'https://rtbx.adgrx.com/impressions/';
   var baseAdapter = Adapter.createNew('adgear');
+  var bidStash = {};
 
   baseAdapter.callBids = function(params) {
     var bidderCode = baseAdapter.getBidderCode();
@@ -33,12 +34,12 @@ var AdGearAdapter = function AdGearAdapter() {
 
           var cuid = utils.getUniqueIdentifierStr();
           url = utils.tryAppendQueryString(url, 'AG_CBID', encodeURIComponent(cuid));
-          bidmanager.pbCallbackMap[cuid] = {
+          bidStash[cuid] = {
             'bidObj': bid,
             'url': url,
             'start': new Date().getTime()
           };
-          bids.push(bidmanager.pbCallbackMap[cuid]);
+          bids.push(bidStash[cuid]);
         });
 
       }
@@ -69,14 +70,14 @@ var AdGearAdapter = function AdGearAdapter() {
    */
   pbjs.handleAdGearCallback = function(bidResponse) {
     if (bidResponse && bidResponse.cbid) {
-      var bidStash = bidmanager.getPlacementIdByCBIdentifer(bidResponse.cbid);
+      var bStash = bidStash[bidResponse.cbid];
       var bid = null;
 
-      if (bidStash) {
+      if (bStash) {
 
-        var bidObj = bidStash.bidObj;
+        var bidObj = bStash.bidObj;
         var timelapsed = new Date().getTime();
-        timelapsed = timelapsed - bidStash.start;
+        timelapsed = timelapsed - bStash.start;
 
         if (bidResponse.width && bidResponse.height && bidResponse.html) {
 
@@ -94,7 +95,7 @@ var AdGearAdapter = function AdGearAdapter() {
 
           bidmanager.addBidResponse(bidObj.placementCode, bid);
 
-          utils.logMessage('[AdGear] Registered bidresponse from htag URL ' + bidStash.url +
+          utils.logMessage('[AdGear] Registered bidresponse from htag URL ' + bStash.url +
                            ' (time: ' + String(timelapsed) + ')');
           utils.logMessage('[AdGear] ======> BID placementCode: ' + bidObj.placementCode +
                            ' CPM: ' + String(bid.cpm) + ' ' + bid.currency +
@@ -106,11 +107,15 @@ var AdGearAdapter = function AdGearAdapter() {
           bid.bidderCode = bidObj.bidder;
           bidmanager.addBidResponse(bidObj.placementCode, bid);
 
-          utils.logMessage('[AdGear] Registered BLANK bidresponse from htag URL ' + bidStash.url +
+          utils.logMessage('[AdGear] Registered BLANK bidresponse from htag URL ' + bStash.url +
                            ' (time: ' + String(timelapsed) + ')');
           utils.logMessage('[AdGear] ======> NOBID placementCode: ' + bidObj.placementCode);
 
         }
+
+        // We should no longer need this stashed object, so drop reference:
+        bidStash[bidResponse.cbid] = null;
+
       }
     }
   };
`

mkendall07 commented:
@bmilekic 

Yes that's fine as the approach. One thing you can do is get the identifier off the bid instead of generating a new one `params.bids[0].bidderRequestId` <= all bids have this identifier and it's the same.
`params.bids[0].bidId` <= all bids have this and it's unique per bid. I think since your grouping requests that `bidderRequestId` makes sense here. Thanks

mkendall07 commented:
@bmilekic any update here? Would like to get this merged soon as we will have another release coming out shortly.

bmilekic commented:
@mkendall07 on it.

bmilekic commented:
@mkendall07 I'm not sure that bidderRequestId is more appropriate than bidId here since we currently do a separate HTTP request per format. I left it as is as I've already tested with it and would suggest deferring changing the unique ID used as the stash index until a future version, if you don't mind.

bmilekic commented:
Rebased and re-pushed. Please do let me know if further changes are required.

bmilekic commented:
@mkendall07 Not sure if this is related, but thought I'd bring it up: while testing manually with integrationExamples/gpt/pbjs_example_gpt.html I run into a sporadic problem. With pbjs_debug=true I am able to see what looks like a timing related issue that appears to occur sporadically (about 50% of reloads)...

When my adgear 300x250 test add shows up successfully I see this sequence in js console:

```
INFO: Invoking pbjs.renderAd : params :  [document, "358cc5625941d4"]
MESSAGE: Calling renderAd with adId :358cc5625941d4
MESSAGE: Emitting event for: bidWon
MESSAGE: Prebid auction cleared
```

When it wins (according to js console logs with pbjs_debug=true, I know it won), but shows up as a blank, I see this sequence in js console:

```
INFO: Invoking pbjs.renderAd : params :  [document, "30d5775c32317a"]
MESSAGE: Calling renderAd with adId :30d5775c32317a
MESSAGE: Emitting event for: bidWon
ERROR: Error trying to write ad. No ad for bid response id: 30d5775c32317a 
MESSAGE: Prebid auction cleared
```

Any clues as to what could be causing renderAd to fail to find the bid response object identified to renderAd? And in particular if it is related to my adapter implementation or something else I tripped on?

mkendall07 commented:
@bmilekic 
Do you have a test file to replicate the behavior? I don't see any issues looking at the code.

bmilekic commented:
@mkendall07 Sure. View source this page: http://bmilekic.scratch.bloomdigital.com/prebid/dev/test2.html?pbjs_debug=true  -- note you won't see ads from AdGear here but if you send me your IP I can get you to occasionally start to get back a 300x250 [winning] bid, at which point you should be able to see the issue replicated in JS console.

mkendall07 commented:
@bmilekic 
I looked at your test page and noticed the error happening. What looks like the problem is that you have multiple HTTP calls being made for the same bid request. So for example, you had 2 bid requests on the page but 4 HTTP calls are being made to `rtbx.adgrx.com`
![image](https://cloud.githubusercontent.com/assets/1870166/17106967/4e42a66a-525b-11e6-8381-13754477063a.png)

Then, we receive the bid response, the `bid.adId` is the same and the nobid response overwrites the good bid response. 

Please check the adapter and make sure you are making single request for each bid. Also please include unit tests for your bidder. Thank you.
